Viewer workload

valschneider opened the pull request:
This workload uses the action.VIEW intent. It will use the
most appropriate package for a given URI, e.g. Gmaps app for
a Gmaps location url. If the device doesn't have Google services,
it will open Gmaps in the browser, and so on.

bjackman commented on libs/utils/android/system.py:
```diff
@@ -153,6 +153,26 @@ def start_action(target, action, action_args=''):
+        arguments = '-d {}'.format(uri)
+
+        if force_new:
+            # Flag is ACTIVITY_NEW_TASK & ACTIVITY_CLEAR_TASK
```
Link to docs/code?

bjackman commented on libs/utils/android/workload.py:
```diff
@@ -62,7 +62,7 @@ def _check_availables(cls, test_env):
         for sc in Workload.__subclasses__():
             _log.debug('Checking workload [%s]...', sc.__name__)
-            if sc.package in cls._packages:
+            if sc.package in cls._packages or sc.package == '':
```
Can it just be `None`? Empty string seems a bit random

valschneider replied:
I had a talk with @derkling about that. Empty string was suggested so that it is obvious to the developer that his workload requires no package (arguably that could also apply to None)

derkling replied:
We can avoid this check by initialising `cls._packages` to always include the empty string.

bjackman replied:
> We can avoid this check by initialising cls._packages to always include the empty string.

I much much prefer the explicit check.

bjackman commented on libs/utils/android/workloads/viewer.py:
```diff
@@ -0,0 +1,108 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2015, ARM Limited and contributors.
```
I don't normally check this but happened to notice this time - year is wrong. Dunno if it makes any difference, probably not TBH...

derkling replied:
I think the rule is to always use the current year a file has been created.

bjackman commented on libs/utils/android/workloads/viewer.py:
```diff
@@ -0,0 +1,108 @@
+    def __init__(self, test_env):
+        super(Viewer, self).__init__(test_env)
+        self._log = logging.getLogger('Viewer')
+        self._log.debug('Workload created')
```
no need for this log

bjackman commented on libs/utils/android/workloads/viewer.py:
```diff
@@ -0,0 +1,108 @@
+        Screen.set_orientation(self._target, portrait=True)
+
+        System.gfxinfo_reset(self._target, self.package)
+        sleep(1)
```
Why sleep? Comment would be good.

derkling replied:
If it's related to resetting the GFXInfo, which may well be... we should add a fixup patch where we include this sleep in the above call and remove this ugly sleep from all workloads using gfxinfo.

bjackman commented on ipynb/examples/android/workloads/Android_Viewer.ipynb:
```diff
@@ -0,0 +1,312 @@
+    "    wload.run(out_dir=te.res_dir,\n",
+    "        collect=\"ftrace\",\n",
+    "        uri=\"https://goo.gl/maps/eAbA8ZxPAy32\",\n",
+    "        workload=lambda: script.run())\n",
```
`lambda: script.run()` == `script.run`

bjackman commented on ipynb/examples/android/workloads/Android_Viewer.ipynb:
```diff
@@ -0,0 +1,312 @@
+    "The Viewer workload will simply read an URI and let Android pick the best application to view the item designated by that URI. That item could be a web page, a photo, a pdf, etc. For instance, if given an URL to a Google Maps location, the Google Maps application will be opened at that location. If the device doesn't have Google Play Services (e.g. HiKey960), it will open Google Maps through ...
+    "\n",
+    "The Viewer workload expects a \"workload\" argument that is a function pointer to some test code to execute while the item is being displayed. In this case we'll simply execute a script on the target, see **${LISA_HOME}/libs/util/target_script.py**"
+   ]
```
Can you add a brief explanation of what the `TargetScript` biz is doing? It might be confusing to a new reader.

bjackman commented on libs/utils/android/workloads/viewer.py:
```diff
@@ -0,0 +1,108 @@
+
+        :param workload: Function pointer to the workload to run once item
+        is opened and being viewed.
+        :type workload: func
```
As you suggested offline I think this param is out of place. A "view workload" sounds like it just views something. So better to have a subclass for the gmaps usecase IMO.

valschneider replied:
My initial thoughts were to go with the function pointer at first for quick testing, then do a proper subclassing. But now that I've been using that function argument, I must admit I quite like it. It's really easy to implement a new workload straight out of a notebook.

derkling replied:
Is not a function pointer just like a pure virtual method? I would probably better go for a subclass me too.
It could be over-killing for some simple use-cases... but if we need to integrate more complex use-cases in the future we already have the infrastructure in place. Isnt' it?

bjackman replied:
Yeah I was also thinking about clarity for this workload itself. IMO It's kind of weird to have a "workload" that says "I view something. Also, you have to pass me a 'workload' to run while I'm viewing things". What if you just want to play a video or get a trace for an image decode?

My first thought was it should be possible to set `workload=None`, then I realised setting this parameter is probably a special enough case that it should be its own class.

valschneider replied:
The thing is, viewing an image or playing a video requires some additional logic anyway: you might want to make sure the image decoding is over, same for video playback (and you may even have to start the video itself). So all this workload should be seen as: "I open apps, but you have to tell me what to do once it's loaded". That could be nothing, that could be swiping, clicking, pressing buttons, etc.

The current setup also makes it really easy to quickly sketch up a workload and tweak it on the go. Having a subclass for each and every workload is contrary to that. I agree that some complex/specific uses of the Viewer could be subclasses, but I don't think that should apply to all of them.

derkling replied:
It does not seems so much more code than what you have in your Notebook.

I think we should be able to defined the class in the notebook itself, did not tried in a Notebook but something like that should work:

```python
class GMapsSwipe(Viewer):

    def __init__(self, test_env):
        super(GMapsSwipe, self).__init__(test_env)
        # Create script
        self._script = TargetScript(self._te, "gmaps_swiper.sh")
        # Define commands to execute during experiment
        for i in range(2):
            System.hswipe(self._script, 40, 60, 100, False)
            script.append('sleep 1')
            System.vswipe(self._script, 40, 60, 100, True)
            script.append('sleep 1')
            System.hswipe(self._script, 40, 60, 100, True)
            script.append('sleep 1')
            System.vswipe(self._script, 40, 60, 100, False)
            script.append('sleep 1')
        self._script.push()

    def workload(self):
        self._script.run()

def experiment():
    # Configure governor
    target.cpufreq.set_all_governors('sched')

    # Create the workload and run it
    wload = GMapsSwipe()
    wload.run(out_dir=te.res_dir,
        collect="ftrace",
        uri="https://goo.gl/maps/eAbA8ZxPAy32")
        
    # Dump platform descriptor
    te.platform_dump(te.res_dir)
```

Isn't it?

In that case I would say we have pretty much the same flexibility than using a function pointer... but we also have a structure which allows to easily deploy the new class under workloads and merge it in LISA.

valschneider replied:
Yes that would be fine to me, but this seems to cause problems with the Workload.getInstance() method (Viewer's subclasses are seen as 'Viewer' themselves). I'll try to look into it.

derkling commented on libs/utils/android/system.py:
```diff
@@ -153,6 +153,26 @@ def start_action(target, action, action_args=''):
     @staticmethod
+    def view_uri(target, uri, force_new=True):
+        """
+        Start a view activity by specifying in URI
```
"by specifying a URI" ?

derkling commented on libs/utils/android/system.py:
```diff
@@ -153,6 +153,26 @@ def start_action(target, action, action_args=''):
+
+        if force_new:
+            # Flag is ACTIVITY_NEW_TASK & ACTIVITY_CLEAR_TASK
+            arguments = '{} -f 0x10008000'.format(arguments)
```
Can we defined this ugly hex as a constant value with a meaningful name at the top of the class?

derkling commented on libs/utils/android/workloads/viewer.py:
```diff
@@ -0,0 +1,108 @@
+        Screen.unlock(self._target)
+
+        # Force screen in PORTRAIT  mode
+        Screen.set_orientation(self._target, portrait=True)
```
Are we sure all possible workload will be happy of a portrait mode?
In case, can it make sense to expose this as a parameter?

derkling commented on libs/utils/android/workloads/viewer.py:
```diff
@@ -0,0 +1,108 @@
+        System.view_uri(self._target, uri)
+
+        # Let the activity start
+        sleep(3)
```
Same here... I would hide the sleep in the above call in such a way to make it kind-of a sync call.
In the future we can make it more robust by checking (maybe via logcat) if the activity has been really started... but that's definitively for another possible future improvement.

valschneider commented:
Most comments should be addressed. I've modified the Workload class so that (subclasses of)+ Workload subclasses are also taken into account.

I've also added pre_viewing(), viewing() and post_viewing() methods to the Viewer class for better tweaking.

Once this is merged I will update gmaps/youtube to use it, and see if some simple workloads could be added (e.g. web page browsing).

derkling commented on libs/utils/android/workload.py:
```diff
@@ -48,6 +48,14 @@ def _adb(self, cmd):
         return 'adb -s {} {}'.format(self._target.adb_name, cmd)
 
     @classmethod
+    def _subclasses(cls):
```
That's cool! Thanks!

derkling commented on ipynb/examples/android/workloads/Android_Viewer.ipynb:
```diff
@@ -0,0 +1,320 @@
+    "# Import support for Android devices\n",
+    "from android import Screen, Workload, System\n",
+    "from target_script import TargetScript\n",
+    "from viewer import Viewer\n",
```
Maybe we can avoid this by moving the definition of the `Viewer` class beside that of `Workload`?
That way it should also be possible to export `Viewer` (which I would also rename into `ViewerWorkload`) from the same `__init__.py` which already exports `Worklaod`, i.e. `libs/utils/android/__init__.py`.

In that case here it should be possible to have just:
```python
from android import Screen, System, ViewerWorkload
```

Not sure will work... but perhaps worth to try ;-)

valschneider commented on libs/utils/android/system.py:
```diff
@@ -21,9 +21,14 @@
 GET_FRAMESTATS_CMD = 'shell dumpsys gfxinfo {} > {}'
 
+# See https://developer.android.com/reference/android/content/Intent.html#setFlags(int)
+FLAG_ACTIVITY_NEW_TASK = 268435456
```
I didn't know how python would output hexes so I went with dead stupid integers, I'll fix that

valschneider commented on libs/utils/android/workloads/viewer.py:
```diff
@@ -0,0 +1,112 @@
+        # Set of output data reported by the viewer
+        self.db_file = None
+
+    def pre_viewing(self):
```
Yeah it's mostly to surround the tracing calls. The pre-viewing is there mostly for video playback, as sometimes the VIEW intent is not enough and you need to press play or something before properly "viewing" the video. Since the `viewing()` method is there for the actual workload (swiping, tapping, waiting...), it's then obvious that `pre_viewing()` and `post_tracing()` are executed just before/after the workload itself. `setup/teardown` works fine too.

valschneider replied:
Also for your below comment on the gfxinfo_reset, I can't reply directly to it so I'll do it here: as there are other workloads that use gfxinfo and then call sleep, I was thinking of writing a separate pull request that removes all of those calls and adds it to the System call itself.

valschneider commented:
Cleaned up things, added docstring, and moved the s/Viewer/ViewerWorkload/ class to **utils/android**. _check_availables() still sees ViewerWorkload as an available workload although it should be a purely abstract workload, so this might be one more thing to clean up.

valschneider commented:
@derkling Jankbench still works in a Notebook and in the command-line benchmarks, so I *think* I haven't broken anything Workload-related. Do you think you can have another look at this PR ? Thanks.
